<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">

<!--Converted with LaTeX2HTML 2002-2-1 (1.71)
original version by:  Nikos Drakos, CBLU, University of Leeds
* revised and updated by:  Marcus Hennecke, Ross Moore, Herb Swan
* with significant contributions from:
  Jens Lippmann, Marek Rouchal, Martin Wilck and others -->
<HTML>
<HEAD>
<TITLE>AVESTRUC</TITLE>
<META NAME="description" CONTENT="AVESTRUC">
<META NAME="keywords" CONTENT="stamp">
<META NAME="resource-type" CONTENT="document">
<META NAME="distribution" CONTENT="global">

<META NAME="Generator" CONTENT="LaTeX2HTML v2002-2-1">
<META HTTP-EQUIV="Content-Style-Type" CONTENT="text/css">

<LINK REL="STYLESHEET" HREF="stamp.css">

<LINK REL="next" HREF="node52.html">
<LINK REL="previous" HREF="node50.html">
<LINK REL="up" HREF="node40.html">
<LINK REL="next" HREF="node52.html">
</HEAD>

<BODY >
<!--Navigation Panel-->
<A NAME="tex2html794"
  HREF="node52.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="file:/usr/share/latex2html/icons/next.png"></A> 
<A NAME="tex2html790"
  HREF="node40.html">
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="file:/usr/share/latex2html/icons/up.png"></A> 
<A NAME="tex2html784"
  HREF="node50.html">
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="file:/usr/share/latex2html/icons/prev.png"></A> 
<A NAME="tex2html792"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="file:/usr/share/latex2html/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html795"
  HREF="node52.html">GSTAMP</A>
<B> Up:</B> <A NAME="tex2html791"
  HREF="node40.html">Summary of parameters for</A>
<B> Previous:</B> <A NAME="tex2html785"
  HREF="node50.html">MERGESTAMP</A>
 &nbsp; <B>  <A NAME="tex2html793"
  HREF="node1.html">Contents</A></B> 
<BR>
<BR>
<!--End of Navigation Panel-->

<H2><A NAME="SECTION005211000000000000000">
AVESTRUC</A>
</H2>

<P>
For various reasons, it is often useful to derive `average`
    structures (i.e. for homology modelling, molecular replacement
    search objects, etc.).  STAMP output provides an obvious starting
    point for obtaining an average structure.  AVESTRUC reads in a
    STAMP alignment file, and generates another PDB file containing
    averaged coordinates (either as C alpha or as a polyalanine 
    structure).
<BR>    
<BR>
The format is:
<BR>    
<BR>    <PRE>
    avestruc -f &lt;STAMP alignment file&gt; 
    [ -polyA -c &lt;STAMP char&gt; -t &lt;threshold&gt; -w &lt;window&gt; -aligned ]
</PRE>

<P>
`-f' specifies the file to be considered.  Note that this MUST BE 
    a STAMP alignment file, containing both transformations and a
    sequence alignment.  It will not work on transformation files
    lacking sequence alignment data or STAMP data.
<BR>    
<BR>    `-polyA' generate  polyalanine model, the default is a C alpha
    model
<BR>    
<BR>    `-c <IMG
 WIDTH="19" HEIGHT="33" ALIGN="MIDDLE" BORDER="0"
 SRC="img46.png"
 ALT="$&lt;$">STAMP char<IMG
 WIDTH="19" HEIGHT="33" ALIGN="MIDDLE" BORDER="0"
 SRC="img47.png"
 ALT="$&gt;$">' `-t <IMG
 WIDTH="19" HEIGHT="33" ALIGN="MIDDLE" BORDER="0"
 SRC="img46.png"
 ALT="$&lt;$">threshold<IMG
 WIDTH="19" HEIGHT="33" ALIGN="MIDDLE" BORDER="0"
 SRC="img47.png"
 ALT="$&gt;$">' `-w <IMG
 WIDTH="19" HEIGHT="33" ALIGN="MIDDLE" BORDER="0"
 SRC="img46.png"
 ALT="$&lt;$">window<IMG
 WIDTH="19" HEIGHT="33" ALIGN="MIDDLE" BORDER="0"
 SRC="img47.png"
 ALT="$&gt;$">'
<BR>
these three parameters tell the program how to define structurally
    equivalent residues.  `STAMP char' is the label of the STAMP field
    specified by the `#' character  in the alignment file.
    `threshold' is the minimum (or maximum in the case of RMS deviation)
    value of the specified STAMP parameter tolerated, and `window' tells
    the minimum number of residues over which this must be true for 
    structural equivalence.  This is less complicated than it sounds.
<BR>    
<BR>
The default is as described in [<A
 HREF="node60.html#rb92b">1</A>]:
<BR>
STAMP char = `G' (i.e. <!-- MATH
 $P_{ij}{\prime}$
 -->
<IMG
 WIDTH="33" HEIGHT="35" ALIGN="MIDDLE" BORDER="0"
 SRC="img50.png"
 ALT="$P_{ij}{\prime}$">)
<BR>
threshold  = 6.0
<BR>
window     = 3 
<BR>    (i.e. stretches of three or more residues having <!-- MATH
 $P_{ij}{\prime} > 6.0$
 -->
<IMG
 WIDTH="82" HEIGHT="35" ALIGN="MIDDLE" BORDER="0"
 SRC="img62.png"
 ALT="$P_{ij}{\prime} &gt; 6.0$">
    are considered equivalent)
<BR>    
<BR>    `-aligned' this flag will generate an averaged position for all positions
    structures are present at a position (i.e. positions not containing any 
    gaps are deamed equivalent).   The temperature factor will then distinguish
    between genuine structural equivalences and fortitously aligned residues.
<BR>    
<BR>    `-ident' `-cons'  these flags will name residues either as a single amino acid
    type (ident) or a conserved type (cons) according to the sequence alignment.
    See the appropriate sections in the preceding chapter for a further explanation.

<P>
<HR>
<!--Navigation Panel-->
<A NAME="tex2html794"
  HREF="node52.html">
<IMG WIDTH="37" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="next"
 SRC="file:/usr/share/latex2html/icons/next.png"></A> 
<A NAME="tex2html790"
  HREF="node40.html">
<IMG WIDTH="26" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="up"
 SRC="file:/usr/share/latex2html/icons/up.png"></A> 
<A NAME="tex2html784"
  HREF="node50.html">
<IMG WIDTH="63" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="previous"
 SRC="file:/usr/share/latex2html/icons/prev.png"></A> 
<A NAME="tex2html792"
  HREF="node1.html">
<IMG WIDTH="65" HEIGHT="24" ALIGN="BOTTOM" BORDER="0" ALT="contents"
 SRC="file:/usr/share/latex2html/icons/contents.png"></A>  
<BR>
<B> Next:</B> <A NAME="tex2html795"
  HREF="node52.html">GSTAMP</A>
<B> Up:</B> <A NAME="tex2html791"
  HREF="node40.html">Summary of parameters for</A>
<B> Previous:</B> <A NAME="tex2html785"
  HREF="node50.html">MERGESTAMP</A>
 &nbsp; <B>  <A NAME="tex2html793"
  HREF="node1.html">Contents</A></B> 
<!--End of Navigation Panel-->
<ADDRESS>
gjbarton
2010-05-10
</ADDRESS>
</BODY>
</HTML>
